### DISCUSSION TITLE
Add GPIO binary sensors (reed-switches), ultrasonic sensors, and HX711 functionality

### DISCUSSION THREAD

**Discussion by CeeWeasel** [ESPresense/ESPresense#103] [Ideas]
These are the last few unimplemented features that I use with ESPHome, and it would be nice to be able to consolidate my devices as much as possible.

I understand that it would be more difficult to implement the HX711 functionality, as you'd need to replicate the calibrate_linear filter and create a UI for entering the information, so consider that a "wishlist" item.

Regarding UI, I would almost think it would be easier to just list all the available pins on the board, and just have a drop-down for each pin where a user could select an appropriate device to connect to it. The list would be filtered based on the type of pin so that, for example, touch sensors would only be listed on pins that have that functionality.

After writing that out, I'm starting to see how much work would actually be involved in something like that, so again, consider this a wishlist.

**Comment by corgan2222** [comment]
I have some reed-switches and ultrasonic sensors here. So I could look into it. 
But I don't have a HX711 Sensor (and have no need for one). If you like to [send one to me](https://www.amazon.de/hz/wishlist/ls/30QURXYJ6P1FM?ref_=wl_share), I could implement it.

> Regarding UI, I would almost think it would be easier to just list all the available pins on the board, and just have a drop-down for each pin where a user could select an appropriate device to connect to it. The list would be filtered based on the type of pin so that, for example, touch sensors would only be listed on pins that have that functionality.

The idea is great and this is the way Tasmota handles it. 
But there are several different ESP32 Boards on the market which have different Pin Layouts. This task alone would really time-consuming, only for coding. Testing is another story.

**Comment by DTTerastar** [comment]
I think we should all remember that these boards cost $4 and positioning the device for optimum bluetooth reception is not the same location as all the other sensors...

**Reply by CeeWeasel** [reply]
You do make a good point. The only thing I can say in contention is that reception doesn't matter if the 2.4Ghz band is full of $4 devices all serving a single purpose. If I already have a dozen ESP32s in use for other tasks, it would be much easier to add ESPresense functionality to them, instead of filling up the network and wifi band with another dozen to only use for tracking.

**Comment by LordNex** [comment]
I was looking for this exact feature really. I already have some old ESP8266 devices in ESPHome with various environmental sensors. I just bought some ESPWROOM32 boards that I was planning on replacing with and was also looking into Internal House Presence Detection. It would be nice if I could create both the sensors config alongside ESPresance so that I can incorporate them all into one project box with a single USB cable. I hate to have to dedicate the ESP32s t just presence and still have to have a separate device in each room to handle environmental sensors.

So if anyone has had any success with this, please let me know. Personally, I try to keep as many of my sensors on the I2C pins since they can be daisy chain them together
